ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+            node_diffuse->eta = node_specular->eta;
+        }
+        break;
+        case CElementBSDF::DIELECTRIC:
+        case CElementBSDF::THINDIELECTRIC:
+        case CElementBSDF::ROUGHDIELECTRIC:
+        {
+            nextSym = ir->allocNode<IR::CBSDFCombinerNode>(IR::CBSDFCombinerNode::ET_FRESNEL_BLEND);
+            nextSym->children.count = 2u;
+
+            const float eta = current->dielectric.intIOR/current->dielectric.extIOR;
+            _IRR_DEBUG_BREAK_IF(eta==1.f);
+            if (eta==1.f)
+                os::Printer::log("WARNING: Dielectric with IoR=1.0!", current->id, ELL_ERROR);
+
+            auto& refl = nextSym->children[0];
+            refl = ir->allocNode<IR::CMicrofacetSpecularBSDFNode>();
+            auto& trans = nextSym->children[1];
+            trans = ir->allocNode<IR::CMicrofacetSpecularBSDFNode>();
+
+            auto* node_refl = static_cast<IR::CMicrofacetSpecularBSDFNode*>(refl);
+            auto* node_trans = static_cast<IR::CMicrofacetSpecularBSDFNode*>(trans);
+
+            node_refl->scatteringMode = IR::CMicrofacetSpecularBSDFNode::ESM_REFLECT;
+            node_refl->shadowing = IR::CMicrofacetSpecularBSDFNode::EST_SMITH;
+            node_refl->eta = IR::INode::color_t(eta);
+            if (currType == CElementBSDF::ROUGHDIELECTRIC)
+            {
+                node_refl->ndf = ndfMap[current->dielectric.distribution];
+                getFloatOrTexture(current->dielectric.alphaU, node_refl->alpha_u);
+                if (node_refl->ndf == IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY)
+                    getFloatOrTexture(current->dielectric.alphaV, node_refl->alpha_v);
+                else
+                    node_refl->alpha_v = node_refl->alpha_u;
+            }
+            else
+            {
+                node_refl->setSmooth();
+            }
+            node_trans->scatteringMode = IR::CMicrofacetSpecularBSDFNode::ESM_TRANSMIT;
+            node_trans->ndf = node_refl->ndf;
+            node_trans->shadowing = node_refl->shadowing;
+            node_trans->alpha_u = node_refl->alpha_u;
+            node_trans->alpha_v = node_refl->alpha_v;
+            node_trans->eta = node_refl->eta;
+
+            thin = (currType == CElementBSDF::THINDIELECTRIC);
+        }
+        break;
+        case CElementBSDF::BUMPMAP:
+        {
+            nextSym = ir->allocNode<IR::CGeomModifierNode>(IR::CGeomModifierNode::ET_DERIVATIVE);
+            nextSym->children.count = 1u;
+
+            auto* node = static_cast<IR::CGeomModifierNode*>(nextSym);
+            node->source = IR::CGeomModifierNode::ESRC_TEXTURE;
+            auto bm = getTexture(current->bumpmap.texture);
+
+            auto& img = std::get<0>(bm)->getCreationParameters().image;
+            auto smplr = std::get<1>(bm);
+
+            std::tie(node->texture.image, node->texture.sampler, node->texture.scale) = std::make_tuple(createDerivMap(img.get(), smplr.get()), std::move(smplr), std::get<2>(bm));
```

Comment:
is the derivative map cached, or do you create a new one every single time?

Reply:
hm, seems like created every time. Thanks for noticing... Now i think maybe this unused map in frontend was meant to be derivative map cache, but never finished.. Anyway this is a TODO